### PR TITLE
Implementation of Map with LRU algorithm.

### LINKED ISSUE #160: Proposal: LruMap
An LruMap is an implementation of Map that removes the least-recently-used item once a threshold (capacity) is reached. Items are promoted to the end (most-recently-used) whenever they are accessed.

(This combined with a quiver.async.MapCache creates a clean LruCache)

### PR DESCRIPTION
**Pull request by matanlurey**
Closes #160.

I considered adding 'weight', but will address that in a separate PR.

### CODE REVIEW

**Comment by justinfagnani** [comment]
New year, new ping :)

**Comment by yjbanov** [comment]
Matan, would you mind if I took this PR over and cleaned it up?

**Comment by matanlurey** [comment]
THE SUSPENSE IS KILLING ME. But really I have no excuse, the work is done on my hometop...

Let me update.

File: lib/collection.dart [hunk 1]
```diff
@@ -21,6 +21,7 @@ import 'dart:collection';
 import 'dart:math';
 
 part 'src/collection/bimap.dart';
+part 'src/collection/lru_map.dart';
```
Comment (matanlurey) [inline lib/collection.dart]: Done. See below in map_cache.dart.

**Comment by matanlurey** [comment]
Ladies and gentlemen.... It's finally here!

(Please do not merge yet - the new implementation is more complex than the old - and I want to add more test cases) ... PTAL otherwise.

File: lib/src/collection/lru_map.dart [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+ */
+abstract class LruMap<K, V> implements Map<K, V> {
+  /**
+   * Creates a [LruMap] instance with the default implementation.
+   */
+  factory LruMap({int maximumSize}) = LinkedLruHashMap;
+
+  /**
+   * Maximum size of the [Map]. If [length] exceeds this value at any time, the
+   * oldest entries are removed beneath the current [length].
```
Comment (yjbanov) [inline lib/src/collection/lru_map.dart]: s/the oldest entries/entries accessed the earliest/ maybe? "the oldest" makes it sound "added a long time ago".

It's probably my ESL, but the phrase "beneath the current length" sounds a bit fuzzy to me.

File: lib/src/collection/lru_map.dart [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+   */
+  int maximumSize;
+}
+
+/**
+ * Simple implementation of a linked-list entry that contains a [key] & [value].
```
Comment (yjbanov) [inline lib/src/collection/lru_map.dart]: Characters such as `&` have a special place in my brain. I'd just spell out "and a value".

File: lib/src/collection/lru_map.dart [hunk 4]
```diff
@@ -0,0 +1,286 @@
+part of quiver.collection;
+
+/**
+ * An implementation of a [Map] which has a maximum size and uses a Least
+ * Recently Used algorithm to remove items from the [Map] when the [maximumSize]
```
Comment (yjbanov) [inline lib/src/collection/lru_map.dart]: s/Least Recently Used/[Least Recently Used](http://en.wikipedia.org/wiki/Cache_algorithms#LRU)/

File: lib/src/collection/lru_map.dart [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+  V value;
+
+  _LinkedEntry<K, V> next;
+  _LinkedEntry<K, V> previous;
+
+  _LinkedEntry([this.key, this.value]);
+}
+
+/**
+ * An [Iterator]/[Iterable] combo for navigating through [_LinkedEntry].
```
Comment (yjbanov) [inline lib/src/collection/lru_map.dart]: ditto re character `/`

File: lib/src/collection/lru_map.dart [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+/**
+ * Simple implementation of a linked-list entry that contains a [key] & [value].
+ */
+class _LinkedEntry<K, V> {
+  K key;
+  V value;
+
+  _LinkedEntry<K, V> next;
+  _LinkedEntry<K, V> previous;
```
Comment (yjbanov) [inline lib/src/collection/lru_map.dart]: Does "linked" imply "doubly linked" or "singly linked". Perhaps we want to be more explicit? Unfortunately the Dart SDK does not provide any guidance. There's `LinkedList` that's actually doubly-linked and `DoubleLinkedQueue`. Rhetorical question, why does Dart API say "*double*-linked"?

File: lib/src/collection/lru_map.dart [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+  void set maximumSize(int maximumSize) {
+    if (maximumSize == null) throw new ArgumentError.notNull('maximumSize');
+    while (length > maximumSize) {
+      _removeLru();
+    }
+    _maximumSize = maximumSize;
+  }
+
+  /**
+   * Look up the value of [key], or add a new value if it isn't there.
```
Comment (yjbanov) [inline lib/src/collection/lru_map.dart]: s/value of/value associated with/

File: lib/src/collection/lru_map.dart [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+   *
+   * Returns the value associated to [key], if there is one, and promotes the
+   * key-value pair to the MRU position.
```
Comment (yjbanov) [inline lib/src/collection/lru_map.dart]: The second paragraph seems to repeat a lot of what the first one said. Can we combine the two?

File: lib/src/collection/lru_map.dart [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+  }
+
+  /**
+   * Get the value for a [key] in the [Map]. The [key] will be promoted to the
+   * 'Most Recently Used' position. *NOTE*: Calling [] inside an iteration over
+   * keys/values is currently unsupported; use [keys] or [values] if you
+   * need information about entries without modifying their position.
+   */
+  @override
+  V operator[] (K key) {
```
Comment (yjbanov) [inline lib/src/collection/lru_map.dart]: https://www.dartlang.org/articles/style-guide/#dont-include-a-space-between-the-declared-name-of-a-method-operator-or-setter-and-its-parameter-list

File: lib/src/collection/lru_map.dart [hunk 10]
```diff
@@ -0,0 +1,286 @@
+part of quiver.collection;
+
+/**
+ * An implementation of a [Map] which has a maximum size and uses a Least
+ * Recently Used algorithm to remove items from the [Map] when the [maximumSize]
+ * is reached and new items are added.
```
Comment (yjbanov) [inline lib/src/collection/lru_map.dart]: Should we mention that using `keys` and `values` does not affect keys' "used" ordering?

File: test/collection/lru_map_test.dart [hunk 11]
```diff
@@ -0,0 +1,101 @@
+// Copyright 2013 Google Inc. All Rights Reserved.
```
Comment (yjbanov) [inline test/collection/lru_map_test.dart]: 2014

File: lib/src/collection/lru_map.dart [hunk 12]
```diff
@@ -0,0 +1,286 @@
+part of quiver.collection;
```
Comment (yjbanov) [inline lib/src/collection/lru_map.dart]: Add license header

File: test/collection/lru_map_test.dart [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+library quiver.collection.lru_map_test;
+
+import 'package:quiver/collection.dart';
+import 'package:unittest/unittest.dart';
+
+void main() {
+  group('LruMap', () {
```
Comment (yjbanov) [inline test/collection/lru_map_test.dart]: Add test the verifies that calling `keys` and `values` multiple times (2 times is enough for the test) does not change the order of entries.

**Comment by yjbanov** [comment]
LGTM w/ some nitpicking at docs and tests.

File: README.md [hunk 14]
```diff
@@ -82,6 +82,8 @@ false.
 
 `listsEqual`, `mapsEqual` and `setsEqual` check collections for equality.
 
+`LruMap` is a map that removes the LRU item when a threshold length is exceeded.
```
Comment (justinfagnani) [inline README.md]: LRU -> least recently used

File: lib/src/cache/map_cache.dart [hunk 15]
```diff
@@ -21,10 +21,18 @@ class MapCache<K, V> implements Cache<K, V> {
...
-   * Creates a new [LocalCache], optionally using [map] as the backing [Map].
+   * Creates a new [MapCache], optionally using [map] as the backing [Map].
    */
   MapCache({Map<K, V> map}) : _map = map != null ? map : new HashMap<K, V>();
 
+  /**
+   * Creates a new [MapCache], using [LruMap] as the backing [Map].
+   * Optionally specify [maximumSize].
+   */
+  factory MapCache.lru({int maximumSize}) {
```
Comment (justinfagnani) [inline lib/src/cache/map_cache.dart]: :beers:

File: lib/src/collection/lru_map.dart (hunk 10 above)
Comment (justinfagnani) [inline lib/src/collection/lru_map.dart]: We should definitely explicitly define what constitutes "use".

File: lib/src/collection/lru_map.dart [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+
+  _LinkedEntry([this.key, this.value]);
+}
+
+/**
+ * An [Iterator]/[Iterable] combo for navigating through [_LinkedEntry].
+ */
+class _LinkedEntryIterator<K, V>
+    extends IterableBase<_LinkedEntry<K, V>>
+    implements Iterator<_LinkedEntry<K, V>> {
```
Comment (justinfagnani) [inline lib/src/collection/lru_map.dart]: Whoa, why does this implement both `Iterable` and `Iterator`? They usually aren't the same thing, since an `Iterator` is stateful, and an `Iterable` can produce many `Iterator`s with independent iteration state.

File: lib/src/collection/lru_map.dart (hunk 6 above)
Comment (justinfagnani) [inline lib/src/collection/lru_map.dart]: Personally, I'm not too concerned about private names. Follow your heart.

File: lib/src/collection/lru_map.dart [hunk 17]
```diff
@@ -0,0 +1,286 @@
...
+  factory LinkedLruHashMap({int maximumSize}) =>
+      new LinkedLruHashMap._fromMap(new HashMap<K, _LinkedEntry<K, V>>(),
+          maximumSize: maximumSize);
+
+  LinkedLruHashMap._fromMap(
+      this._entries, {
+      int maximumSize})
+          // This pattern is used instead of a default value because we want to
+          // be able to respect null values coming in from MapCache.lru.
+          : _maximumSize = maximumSize != null ? maximumSize
```
Comment (justinfagnani) [inline lib/src/collection/lru_map.dart]: `import 'package:quiver/core.dart;`

`: _maximumSize = firstNonNull(maximumSize, _DEFAULT_MAXIMUM_SIZE)`

File: lib/src/collection/lru_map.dart [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+              : _DEFAULT_MAXIMUM_SIZE;
+
+  /**
+   * Iterates over [other], adding all key-value pairs to the map. The MRU/LRU
+   * order is based off of [other.forEach] implementation - the first item would
+   * be the LRU, and the last item, the MRU.
```
Comment (justinfagnani) [inline lib/src/collection/lru_map.dart]: The first item is only the LRU of the new keys, existing keys not overwritten by `other` are more least recently used.

File: lib/src/collection/lru_map.dart [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+   *
+   * Because [maximumSize] is still enforced, this means if [other.length] is
```
Comment (justinfagnani) [inline lib/src/collection/lru_map.dart]: I'd rephrase this a bit, and base it off of the `Map.addAll` docs:

Adds all key-value pairs of `other` to this map.

The operation is equivalent to doing this[key] = value for each key and associated value in other. It iterates over other, which must therefore not change during the iteration.

If a key of `other` is already in this map, its value is overwritten. If the number of unique keys is greater than [maximumSize] then the least recently use keys are evicted. For keys written to by `other`, the least recently user order is determined by `other`'s iteration order.

File: lib/src/collection/lru_map.dart [hunk 20]
```diff
@@ -0,0 +1,286 @@
...
+  }
+
+  /**
+   * Removes the LRU position, shifting the linked list if necessary.
+   */
+  void _removeLru() {
+    // Remove the tail from the internal map.
+    _entries.remove(_tail.key);
+
+    // Remove the tail element itself.
```
Comment (justinfagnani) [inline lib/src/collection/lru_map.dart]: null out the previous tail's previous pointer?

**Comment by justinfagnani** [comment]
I'd like to talk about the Iteratable/Iterable combo object before merging.

File: lib/src/collection/lru_map.dart [hunk 21]
```diff
@@ -0,0 +1,286 @@
...
+    implements Iterator<_LinkedEntry<K, V>> {
+  @override
+  _LinkedEntry<K, V> current;
+
+  _LinkedEntryIterator(_LinkedEntry<K, V> head) {
+    current = new _LinkedEntry<K, V>()..next = head;
+  }
+
+  @override
+  Iterator<_LinkedEntry<K, V>> get iterator => this;
```
Comment (justinfagnani) [inline lib/src/collection/lru_map.dart]: This is definitely broken here.

File: lib/src/collection/lru_map.dart (hunk 6 above)
Comment (matanlurey) [inline lib/src/collection/lru_map.dart]: Dart's own SDK has _Linked entries that are doubly linked. I'll keep as is for simplicity for now.

File: lib/src/collection/lru_map.dart (hunk 12 above)
Comment (matanlurey) [inline lib/src/collection/lru_map.dart]: Done.

File: test/collection/lru_map_test.dart (hunk 13 above)
Comment (matanlurey) [inline test/collection/lru_map_test.dart]: Done.

File: lib/src/collection/lru_map.dart (hunk 21 above)
Comment (matanlurey) [inline lib/src/collection/lru_map.dart]: I'm not super stoked either. Is there a better way? One way I considered was just creating:

```dart
class _SimpleIterable<T> extends IterableBase<T> {
  @override
  final Iterator<T> iterator;

  _SimpleIterable(this.iterator);
}
```

And doing:

```dart
return new _SimpleIterable(new _LinkedEntryIterator(...))
```

(I think that ```_SimpleIterable```, if something like it doesn't already exist, might be a nice candidate for quiver/collections itself)

File: lib/src/collection/lru_map.dart (hunk 20 above)
Comment (matanlurey) [inline lib/src/collection/lru_map.dart]: Curious, why? _tail is effectively out of the list.

File: lib/src/collection/lru_map.dart (hunk 19 above)
Comment (matanlurey) [inline lib/src/collection/lru_map.dart]: Done, this sounds great.

File: lib/src/collection/lru_map.dart (hunk 18 above)
Comment (matanlurey) [inline lib/src/collection/lru_map.dart]: Done.

File: lib/src/collection/lru_map.dart (hunk 17 above)
Comment (matanlurey) [inline lib/src/collection/lru_map.dart]: Nice, done.

File: README.md (hunk 14 above)
Comment (matanlurey) [inline README.md]: Done.

**Comment by matanlurey** [comment]
Ping.

File: lib/src/collection/lru_map.dart (hunk 21 above)
Comment (justinfagnani) [inline lib/src/collection/lru_map.dart]: No, that wouldn't work either. You simply can't tie an instance of `Iterable` to an instance of an `Iterator`. `Iterable.iterator` _must_ return a new instance each time it's called. It's probably a bug that it's a getter.

Comment (justinfagnani) [inline lib/src/collection/lru_map.dart]: You could do something like:

```
typedef Iterator<T> IteratorFactory<T>();

class _SimpleIterable<T> extends IterableBase<T> {
  final IteratorFactory<T>() _newIterator;

  @override
  Iterator<T> get iterator => _newIterator();

  _SimpleIterable(this._newIterator);
}
```

Comment (justinfagnani) [inline lib/src/collection/lru_map.dart]: As for something in Quiver, `GeneratingIterable` would also help:

```
  Iterable<K> get keys => new GeneratingIterable<K>(() => _head, (n) => n.next).map((e) => e.key);
```

Comment (matanlurey) [inline lib/src/collection/lru_map.dart]: Thanks Justin, this worked very nice and clean. Done.

File: lib/src/collection/lru_map.dart [hunk 22]
```diff
@@ -0,0 +1,303 @@
...
+   */
+  Iterable<_LinkedEntry<K, V>> _iterable() {
+    return new GeneratingIterable<_LinkedEntry<K, V>>(
+        () => _head, (n) => n.next);
+  }
+
+  /**
+   * The keys of [this] - in order of MRU to LRU.
+   *
+   * The returned iterable does *not* have efficient `length` or `contains`.
```
Comment (justinfagnani) [inline lib/src/collection/lru_map.dart]: Interesting... it definitely could with a pretty simple subclass of `DelegatingIterable`. Maybe a todo?

**Comment by justinfagnani** [comment]
LGTM! Can you squash the commits?

**Comment by matanlurey** [comment]
Done.

**Comment by justinfagnani** [comment]
:beers:
